File: warmups/search-algorithms/05.quick-sort/main.py
def partition(arr: list[int], start: int, end: int) -> int:
  pivot = arr[start]
  left = start
  right = end

  while left < right:
    while left < end and arr[left] <= pivot:
      left += 1
    while right > start and arr[right] > pivot:
      right -= 1

      if left < right:
        arr[left], arr[right] = arr[right], arr[left]

  arr[start], arr[right] = arr[right], arr[start]

  return right


def quick_sort(arr: list[int], start: int, end: int) -> None:
  if start < end:
    partition_index = partition(arr, start, end)
    quick_sort(arr, start, partition_index - 1)
    quick_sort(arr, partition_index + 1, end)


input_array = [12, 5, 9, 4, 10, 43]
print(input_array)

quick_sort(input_array, 0, len(input_array) - 1)
print(input_array)

File: warmups/search-algorithms/05.quick-sort/main.js
const inputArray = [ 12, 5, 9, 4, 10, 43 ];
console.log(inputArray);

const partition = (arr, start, end) => {
  const pivot = arr[start];
  let left = start;
  let right = end;

  while (left < right) {
    while (arr[left] <= pivot && left < end) {
      left++;
    }
    while (arr[right] > pivot && right > start) {
      right--;
    }
    if (left < right) {
      [arr[left], arr[right]] = [arr[right], arr[left]];
    }
  }

  [arr[start], arr[right]] = [arr[right], arr[start]];

  return right;
}

const quickSort = (arr, start, end) => {
  if (start < end) {
    const partitionIndex = partition(arr, start, end);
    quickSort(arr, start, partitionIndex - 1);
    quickSort(arr, partitionIndex + 1, end);
  }
}

quickSort(inputArray, 0, inputArray.length - 1);
console.log(inputArray);
